examples/chat/src/network.rs

```diff
@@ -0,0 +1,486 @@
+                        return;
+                    }
+                };
+
+                let _ = sender.send(Ok(id));
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum Command {
+    ListenOn {
+        addr: Multiaddr,
+        sender: oneshot::Sender<eyre::Result<()>>,
+    },
+    Dial {
+        peer_addr: Multiaddr,
+        sender: oneshot::Sender<eyre::Result<Option<()>>>,
+    },
+    Subscribe {
+        topic: gossipsub::IdentTopic,
+        sender: oneshot::Sender<eyre::Result<gossipsub::IdentTopic>>,
+    },
+    Unsubscribe {
+        topic: gossipsub::IdentTopic,
+        sender: oneshot::Sender<eyre::Result<gossipsub::IdentTopic>>,
+    },
+    PeerInfo {
+        sender: oneshot::Sender<PeerInfo>,
+    },
+    MeshPeerCount {
+        topic: gossipsub::TopicHash,
+        sender: oneshot::Sender<MeshPeerInfo>,
+    },
+    Publish {
+        topic: gossipsub::TopicHash,
+        data: Vec<u8>,
+        sender: oneshot::Sender<eyre::Result<gossipsub::MessageId>>,
+    },
+}
+
+#[allow(dead_code)] // Info structs for pretty printing
+#[derive(Debug)]
+pub(crate) struct PeerInfo {
+    count: usize,
+    peers: Vec<PeerId>,
+}
+
+#[allow(dead_code)] // Info structs for pretty printing
+#[derive(Debug)]
+pub(crate) struct MeshPeerInfo {
+    count: usize,
+    peers: Vec<PeerId>,
+}
+
+#[derive(Debug)]
+pub(crate) struct DialEntry {
+    pub(crate) sender: oneshot::Sender<eyre::Result<Option<()>>>,
+    pub(crate) dial_state: PendingDial,
+}
+
+#[derive(Debug)]
+pub(crate) enum DialPreference {
+    Direct {
+        peer_id: PeerId,
+    },
+    Dcutr {
+        peer_id: PeerId,
+        relay_peer_id: PeerId,
+    },
+}
+
+#[derive(Debug)]
+pub(crate) enum PendingDial {
+    Direct,
+    Dcutr(PendingDialDcutr),
+}
+
+#[derive(Debug)]
+pub(crate) struct PendingDialDcutr {
+    pub(crate) relay_peer_id: PeerId,
+    pub(crate) state: PendingDialDcutrState,
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum PendingDialDcutrState {
+    Initial,
+    RelayConnected,
+}
+
+impl TryFrom<&Multiaddr> for DialPreference {
+    type Error = &'static str;
+
+    fn try_from(value: &Multiaddr) -> Result<Self, Self::Error> {
+        // If there's no p2p-circuit protocol, directly extract peer_id
+        if !value
+            .iter()
+            .any(|protocol| matches!(protocol, multiaddr::Protocol::P2pCircuit))
+        {
+            for protocol in value.iter() {
+                if let multiaddr::Protocol::P2p(peer_id) = protocol {
+                    return Ok(DialPreference::Direct { peer_id });
+                }
+            }
+        }
+
+        // Iterate over the protocols in the MultiAddr
+        let mut iter = value.iter();
+        let mut p2p_circuit_state = false;
+        let mut peer_id = None;
+        let mut relay_peer_id = None;
+
+        while let Some(protocol) = iter.next() {
+            match protocol {
+                multiaddr::Protocol::P2pCircuit => {
```

Comment:
also, to avoid silently handling undefined behaviour, if we already have a `peer_id`, disallow subsequent `P2pCircuit`s